### PR TITLE
JS: Add ModuleImportNode::Range

### PR DESCRIPTION
**Pull request by asger-semmle**
Will conflict with https://github.com/Semmle/ql/pull/934 which is undergoing a long evaluation. Will evaluate this after rebasing on that.

Would very much appreciate it if this could make it into 1.20, which is why I'm opening this ASAP.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll [hunk 1]
```diff
@@ -403,51 +403,77 @@ class ArrayCreationNode extends DataFlow::ValueNode, DataFlow::SourceNode {
...
+  ModuleImportNode() { this = range }
 
   /** Gets the path of the imported module. */
-  string getPath() { result = path }
+  string getPath() { result = range.getPath() }
+}
+
+module ModuleImportNode {
+  /**
+   * Data flow node that refers to an imported module.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: Missing `A`?

> **A** data flow node that refers to an imported module.

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll [hunk 2]
```diff
@@ -403,51 +403,77 @@ class ArrayCreationNode extends DataFlow::ValueNode, DataFlow::SourceNode {
...
+    override string getPath() { result = path }
+  }
 }
 
 /**
- * Gets a (default) import of the module with the given path.
+ * Gets a (default) import of the module with the given path, such as `require("fs")`
+ * or `import * as fs from "fs"`.
+ *
+ * This predicate can be customized by subclassing `ModuleImportNode::Range`.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: Do we use "customized" like this anywhere else? 
You can only add `ModuleImportNode`s to this predicate when subclassing `ModuleImportNode::Range`, so "customize" seems like a stretch.

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll [hunk 3]
```diff
@@ -403,51 +403,77 @@ class ArrayCreationNode extends DataFlow::ValueNode, DataFlow::SourceNode {
...
   /** Gets the path of the imported module. */
-  string getPath() { result = path }
+  string getPath() { result = range.getPath() }
+}
+
+module ModuleImportNode {
+  /**
+   * Data flow node that refers to an imported module.
+   */
+  abstract class Range extends DataFlow::SourceNode {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: Do we need this to extend `DataFlow::SourceNode` itself, or could it extend `DataFlow::SourceNode::Range` instead?

**Comment by xiemaisi** [comment]
LGTM, modulo minor comments. I don't think this needs an evaluation, but if you want to do a quick sanity check I won't stand in your way.

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll (hunk 2 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: Changed to "extended"

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll (hunk 3 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: Extending `SourceNode::Range` will add more dependencies, which could lead to negative recursion or performance issues. I believe there was only a one-way dependency between `ModuleImportNode` and `SourceNode` previously so it seems safer to keep it that way.

**Comment by asger-semmle** [comment]
[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/augustus.ti.semmle.com_1551170665188) with just Xss.ql looks fine. I also checked the profile viewer and they look identical.

**Pull request merged**